### PR TITLE
Use a single file for config store read/write with data stream

### PR DESCRIPTION
**Pull request by tjwatson**

### CODE REVIEW

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#build

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStore.java [hunk 1]
```diff
@@ -189,83 +124,61 @@ public synchronized void unbindConfigurations(Bundle bundle) {
...
+        if (configDatas.isFile()) {
+            ConfigStorageConsumer<ExtendedConfigurationImpl> consumer = new ConfigStorageConsumer<ExtendedConfigurationImpl>() {
+                @Override
+                public ExtendedConfigurationImpl consumeConfigData(String location, Set<String> uniqueVars, Set<ConfigID> references, ConfigurationDictionary dict) {
+                    String pid = (String) dict.get(Constants.SERVICE_PID);
+                    String factoryPid = (String) dict.get(ConfigurationAdmin.SERVICE_FACTORYPID);
+                    return new ExtendedConfigurationImpl(caFactory, location, factoryPid, pid, dict, references, uniqueVars);
+                }
+            };
+            Collection<ExtendedConfigurationImpl> configs = Collections.emptyList();
```
Comment (NottyCode) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStore.java]: could the init of configs to Collections.emptyList be moved into the catch block? On the golden path we are going to end up creating two lists here and I expect the golden path should be over 99% of cases.

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStore.java [hunk 2]
```diff
@@ -189,83 +124,61 @@ public synchronized void unbindConfigurations(Bundle bundle) {
...
+            try {
+                configs = ConfigurationStorageHelper.load(configDatas, consumer);
```
Comment (NottyCode) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStore.java]: Based just on what is happening here it seems that perhaps ConfigurationStorageHelper.load would be more useful if it returned a Map so line 160-162 isn't needed.

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+import com.ibm.ws.config.admin.ExtendedConfiguration;
+import com.ibm.wsspi.kernel.service.utils.OnErrorUtil.OnError;
+import com.ibm.wsspi.kernel.service.utils.SerializableProtectedString;
+
+/**
+ * A helper class for storing and retrieving the persisted config information.
+ *
+ * The format is:
+ *
+ * byte - version, set to 0
```
Comment (NottyCode) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java]: The current code is out of sync with this description. So far all I can say for sure is that after the version an int number of configs is written.

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java [hunk 4]
```diff
@@ -0,0 +1,673 @@
...
+                    String location = load(dis, uniqueVars, references, dict);
+                    result.add(consumer.consumeConfigData(location, uniqueVars, references, dict));
+                }
+                return result;
+            }
+        }
+        return Collections.emptyList();
+    }
+
+    public static String load(DataInputStream dis, Set<String> uniqueVars, Set<ConfigID> references, ConfigurationDictionary dict) throws IOException {
```
Comment (NottyCode) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java]: does this need to be public?

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java [hunk 5]
```diff
@@ -0,0 +1,673 @@
...
+                Dictionary<String, ?> properties = config.getReadOnlyProperties();
+                Set<ConfigID> references = config.getReferences();
+                Set<String> uniqueVars = config.getUniqueVariables();
+                String bundleLocation = config.getBundleLocation();
+                store(dos, properties, bundleLocation, references, uniqueVars);
+            }
+        }
+    }
+
+    public static void store(DataOutputStream dos, Dictionary<String, ?> properties, String bundleLocation,
```
Comment (NottyCode) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java]: does this need to be public?

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java [hunk 6]
```diff
@@ -0,0 +1,673 @@
...
+            for (ConfigID id : references) {
+                count = writeConfigID(dos, id, count, 0, writtenConfigIds);
+            }
+        } else {
+            dos.writeInt(0);
+        }
+        writeMap(dos, toMapOrDictionary(properties));
+    }
+
+    public static void readMap(DataInputStream dis, MapIterable props) throws IOException {
```
Comment (NottyCode) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java]: does this need to be public?

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java [hunk 7]
```diff
@@ -0,0 +1,673 @@
...
+            value = map;
+            readMap(dis, toMapOrDictionary(map));
+        } else {
+            value = readPrimative(dis, type);
+        }
+
+        return value;
+    }
+
+    // TODO how to handle null values?
```
Comment (NottyCode) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java]: does this TODO still apply?

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java [hunk 8]
```diff
@@ -0,0 +1,673 @@
...
+        if (len > 0) {
+            int type = dis.readByte();
+            for (int i = 0; i < len; i++) {
+                result.add(readPrimative(dis, type));
+            }
+        }
+        return result;
+    }
+
+    public static void writeMap(DataOutputStream dos, MapIterable map) throws IOException {
```
Comment (NottyCode) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java]: does this need to be public?

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java [hunk 9]
```diff
@@ -0,0 +1,673 @@
...
+                }
+            } else {
+                writePrimitiveType(dos, obj.getClass());
+                writePrimitive(dos, obj);
+            }
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public static MapIterable toMapOrDictionary(Object obj) {
```
Comment (NottyCode) [inline dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java]: does this need to be public?

**Comment by tjwatson** [comment]
#build

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStore.java (hunk 1 above)
Comment (tjwatson) [reply dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStore.java]: I have reworked this code to use a Map type and the result directly assigns the final configurations map instead.

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStore.java (hunk 2 above)
Comment (tjwatson) [reply dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStore.java]: Done

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java (hunk 3 above)
Comment (tjwatson) [reply dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java]: Updated the comments with the format.  This made me realize that I did not store the version for the Map used for the variable registry cache.  This is now fixed.

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java (hunk 5 above)
Comment (tjwatson) [reply dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java]: No, I made the method package private to allow unit testing.

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java (hunk 6 above)
Comment (tjwatson) [reply dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java]: Yes for usage by the variable registry cache which needs to persist a map with the same types as config.

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java (hunk 7 above)
Comment (tjwatson) [reply dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java]: Opps, I still need to address this.  But I'm not sure we allow for arrays.  It doesn't seem like it.

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java (hunk 8 above)
Comment (tjwatson) [reply dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java]: Yes for usage by the variable registry cache

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java (hunk 9 above)
Comment (tjwatson) [reply dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java]: Yes for usage by the variable registry cache

File: dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java (hunk 7 above)
Comment (tjwatson) [reply dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java]: Fixed this with the last commit to handle null values in arrays and mixed type collections

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#build

**Comment by tjwatson** [comment]
#build

**Pull request merged**